### ISSUE TITLE
SPI::transferBytes should handle unaligned buffers

### ISSUE THREAD

**Issue by piflav** [esp8266/Arduino#3315]
Hi
compiling the very same source on macOS and Windows I get different binaries and while the output on Windows works pefectly the one on macOS crash with exception (9) on Wroom-2 (PRODino Board).
Both  Arduino/esp8266 enviroment are fresh and updated installations.

Investigating I found the crash happens on line 463 of SPI.cpp:

`*fifoPtr = *dataPtr;`

calling void SPIClass::transferByte.
The source reports:
```
/**
 * Note:
 *  in and out need to be aligned to 32Bit
 *  or you get an Fatal exception (9)
 * @param out uint8_t *
 * @param in  uint8_t *
 * @param size uint32_t
 */
```

Is that the problem?
Why does it happen on macOS and not on Windows?

Thanks
Flavio

----------------------------------------------------------------------------

Hi,
I'm having the issue working with PRODIno board
https://www.kmpelectronics.eu/en-us/products/prodinowifi-esp.aspx.
Compiling and uploading the basic example
https://github.com/kmpelectronics/Arduino/blob/master/ProDinoWiFiEsp/src/PRODINoESP8266/examples/WiFiWebRelaySrv/WiFiWebRelaySrv.ino
on Windows works while on Mac I get exception (9) (see attached screenshot).
I asked to kmpelectronics but they cannot support with Mac and esp8266 Arduino libraries.
It's clear something wrong on the toolchain because same source different binaries beetween Mac and Windows. Both enviroments are fresh installations. I also tried installing esp8266 platform rather use the package provide by kmp to be sure to have the last and original toolchain+libraries but same result.

Thanks in advance
Ciao
Flavio

![dump](https://cloud.githubusercontent.com/assets/18332405/26783949/267a3754-49fb-11e7-891d-bf9225e073cb.png)

**Comment by piflav** [comment]
Can you please suggest further investigations?
Thanks
Flavio

**Comment by igrr** [comment]
Do you have the minimal sketch required to reproduce this issue? Please attach it to the first post, along with the exception output decoded using ESPExceptionDecoder tool.

**Comment by piflav** [comment]
Any news?

**Comment by igrr** [comment]
Based on the backtrace, indeed that library is passing a non-aligned pointer to SPI::transferBytes.

I would call this a bug in SPI library, because on other Arduino platforms transferBytes doesn't have such alignment restrictions.

**Comment by d-a-v** [comment]
Fixed by #4925
